Add onSelectionEnd function parameter to CircularSlider

kirkeaton opened the pull request:
This pull request adds an additional callback function for when the slider selection has ended. This may be useful in cases where the value of the slider maybe be used to update a value on a server or database. In these cases one could use this function as opposed to onSelectionChange to avoid updating too often.

davidanaya commented:
@kirkeaton thanks for your PR
Sorry I was not aware of it before and I did a few extra updates in the last days.
Could you please review it and fix the conflicts? Otherwise I will incorporate your suggestion myself when I have some time.

kirkeaton commented:
@davidanaya 
I have just rebased this PR against the latest code base, fixed all of the conflicts, and added my changes to the SingleCircularSlider as well.

davidanaya commented on lib/src/circular_slider_paint.dart:
```diff
@@ -1,34 +1,54 @@
   final int secondarySectors;
-  final Function onSelectionChange;
+  final SelectionChanged<int> onSelectionChange;
+  final Function onSelectionEnd;
```
we could use the same typedef that I created for `onSelectionChange` (see line above)

davidanaya replied:
please note that we should also return the laps now

davidanaya commented on lib/src/circular_slider_paint.dart:
```diff
@@ -124,16 +238,46 @@ class _CircularSliderState extends State<CircularSliderPaint> {
+      widget.onSelectionEnd(newValue, widget.end);
+    } else {
+      widget.onSelectionEnd(widget.init, newValue);
+    }
```
this is a bit more complex now
we could have both handlers selected (click on the selection and dragging moves the whole selection) so we would need to account for that (check lines 243 to 251 in this same file for reference)
also we would need to return also the laps (same lines)

davidanaya commented on lib/src/double_circular_slider.dart:
```diff
@@ -0,0 +1,148 @@
+
+  /// callback function when init and end finish
+  /// (int init, int end) => void
+  final Function onSelectionEnd;
```
please use same type than with `onSelectionChange` above

davidanaya commented on lib/src/single_circular_slider.dart:
```diff
@@ -0,0 +1,143 @@
+
+  /// callback function when init and end finish
+  /// (int init, int end) => void
+  final Function onSelectionEnd;
```
please use the same type as `onSelectionChange` above

davidanaya commented on lib/src/double_circular_slider.dart:
```diff
@@ -0,0 +1,148 @@
+              _end = newEnd;
+            });
+          },
+          onSelectionEnd: (newInit, newEnd) {
```
include laps

davidanaya commented on lib/src/single_circular_slider.dart:
```diff
@@ -0,0 +1,143 @@
+            });
+          },
+          onSelectionEnd: (newInit, newEnd) {
+            if (widget.onSelectionEnd != null) {
```
include laps

kirkeaton commented:
@davidanaya Thanks for the review. I think I've addressed all the changes. I've consolidated the panning logic into a separate function to reduce duplicated code as well.

davidanaya commented on lib/src/circular_slider_paint.dart:
```diff
@@ -94,63 +138,181 @@ class _CircularSliderState extends State<CircularSliderPaint> {
+    }
 
+    // isDoubleHandler but one handler was selected
     if (_isInitHandlerSelected) {
```
I think this is unreachable unless you remove the previous code (lines 261 to 264)

kirkeaton commented:
@davidanaya Thanks, I went ahead and removed that if statement.

kirkeaton commented:
@davidanaya Thank's for approving! I'm unable to merge as I don't have write access to the repository.
